File: Data Structure/Stack/acmicpc.net/9012/Main.js
"use strict";
const inputFilePath = "/sample.txt";
const input = require("fs")
  .readFileSync(__dirname + inputFilePath)
  .toString()
  .trim()
  .split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");

class Stack {
  constructor() {
    this._arr = [];
  }
  ["push"](value) {
    return this._arr.push(value);
  }
  ["pop"]() {
    if (this.isEmpty()) {
      return -1;
    } else {
      return this._arr.pop();
    }
  }
  ["size"]() {
    return this._arr.length;
  }
  ["isEmpty"]() {
    if (this._arr.length === 0) {
      return true;
    } else {
      return false;
    }
  }
  ["top"]() {
    if (this.isEmpty()) {
      return -1;
    } else {
      return this._arr[this._arr.length - 1];
    }
  }
}
let ans = [];
for (let i = 1; i <= input[0]; i++) {
  let parenthesisArr = input[i].split("");
  let stack = new Stack();
  let result = "";
  for (const aParenthesis of parenthesisArr) {
    if (aParenthesis === "(") {
      stack.push("(");
    } else {
      if (!stack.isEmpty()) {
        stack.pop();
      } else {
        result = "NO";
        break;
      }
    }
  }
  if (result === "") {
    if (!stack.isEmpty()) {
      ans.push("NO");
    } else {
      ans.push("YES");
    }
  } else {
    ans.push(result);
  }
}

console.log(ans.join("\n"));

File: Data Structure/Stack/acmicpc.net/9012/Main.py
import sys

# input = sys.stdin.readline
file = open("./01 ing/acmicpc.net/sample.txt", "r")
input = file.readline

def inner(ps):
    left = []
    for item in ps:
        if(item == "("):
            left.append("(")
        else:
            if(len(left) == 0):
                print("NO")
                return
            else:
                left.pop()
    if(len(left) ==0):
        print("YES")
    else:
        print("NO")


n = int(input())
for _ in range(n):
    ps = list(input().split()[0])
    inner(ps)
